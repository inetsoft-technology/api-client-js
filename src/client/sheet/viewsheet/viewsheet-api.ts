/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Sheet } from "../sheet-api";

/**
 * ViewsheetAPI provides operations for accessing viewsheets.
 */
export interface ViewsheetApi {
   /**
    * Gets the list of available viewsheets.
    *
    * @return the list of viewsheets.
    */
   getViewsheets(): Promise<Sheet[]>;

   /**
    * Gets the list of opened viewsheets.
    *
    * @return the list of opened viewsheets.
    */
   getOpenedViewsheets(): Promise<Sheet[]>;

   /**
    * Opens a viewsheet.
    *
    * @param assetId    the asset identifier of the viewsheet to open.
    * @param parameters a map of parameters to use when opening the viewsheet.
    *
    * @return the opened viewsheet information.
    */
   openViewsheet(assetId: string, parameters: {[name: string]: any}): Promise<Sheet>;

   /**
    * Gets information about an opened viewsheet.
    *
    * @param id the runtime identifier of the opened viewsheet.
    *
    * @return the opened viewsheet information.
    */
   getOpenedViewsheet(id: string): Promise<Sheet>;

   /**
    * Closes an opened viewsheet.
    *
    * @param id the runtime identifier of the opened viewsheet.
    */
   closeViewsheet(id: string): Promise<void>;

   /**
    * Sends an opened viewsheet in an email message.
    *
    * @param id         the runtime identifier of the opened viewsheet.
    * @param format     the file format for the attached viewsheet.
    * @param recipients the email addresses of the recipients of the message.
    * @param sender     the email address of the sender of the message.
    * @param subject    the subject line of the message.
    * @param message    the body of the message.
    */
   mailViewsheet(id: string, format: string, recipients: string[], sender: string, subject: string,
                 message: string): Promise<void>;

   /**
    * Exports an opened viewsheet.
    *
    * @param id       the runtime identifier of the opened viewsheet.
    * @param format   the format of the exported viewsheet.
    *
    * @return the binary content of the exported viewsheet.
    */
   exportViewsheet(id: string, format: string): Promise<ArrayBuffer>;
}
