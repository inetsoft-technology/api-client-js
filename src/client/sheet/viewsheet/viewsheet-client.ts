/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AbstractClient } from "../../abstract-client";
import { HttpAdapter } from "../../http-adapter";
import { Sheet } from "../sheet-api";
import { ViewsheetApi } from "./viewsheet-api";

/**
 * @hidden
 */
export class ViewsheetClient extends AbstractClient implements ViewsheetApi {

   constructor(adapter: HttpAdapter<any>, token: () => Promise<string>, baseUrl: string) {
      super(adapter, token, baseUrl);
   }

   public getViewsheets(): Promise<Sheet[]> {
      const url = "api/public/viewsheets";
      return this.doGet(url).then((list: any) => list.sheets);
   }

   public getOpenedViewsheets(): Promise<Sheet[]> {
      const url = "api/public/viewsheets/open";
      return this.doGet(url).then((list: any) => list.sheets);
   }

   public openViewsheet(assetId: string, parameters: { [p: string]: any }): Promise<Sheet> {
      const request = { asset: assetId, parameters };
      const url = "api/public/viewsheets/open";
      return this.doPost(url, request);
   }

   public getOpenedViewsheet(id: string): Promise<Sheet> {
      const url = `api/public/viewsheets/open/${id}`;
      return this.doGet(url);
   }

   public closeViewsheet(id: string): Promise<void> {
      const url = `api/public/viewsheets/open/${id}`;
      return this.doDelete(url);
   }

   public mailViewsheet(id: string, format: string, recipients: string[], sender: string,
                        subject: string, message: string): Promise<void>
   {
      const request = { format, recipients, sender, subject, message };
      const url = `api/public/viewsheets/open/${id}/mail`;
      return this.doPost(url, request);
   }

   public exportViewsheet(id: string, format: string): Promise<ArrayBuffer> {
      const url = `api/public/viewsheets/open/${id}/export/${format}`;
      return this.doGetBinary(url);
   }
}
