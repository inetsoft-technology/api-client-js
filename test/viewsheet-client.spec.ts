/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ViewsheetClient } from "../src/client/sheet/viewsheet/viewsheet-client";

describe("ViewsheetClient", () => {
   let adapter: any;
   let client: ViewsheetClient;

   const sheets = [
      {
         asset: "1^128^__NULL__^Examples/Census",
         path: "Examples/Census",
         label: "Census",
         global: true
      },
      {
         asset: "1^128^__NULL__^Examples/Hurricanes",
         path: "Examples/Hurricanes",
         label: "Hurricanes",
         global: true
      },
      {
         asset: "1^128^__NULL__^Examples/Sales Summary",
         path: "Examples/Sales Summary",
         label: "Sales Summary",
         global: true
      }
   ];

   const openSheets = [
      {
         asset: "1^128^__NULL__^Examples/Census",
         path: "Examples/Census",
         label: "Census",
         global: true,
         identifier: "id1"
      },
      {
         asset: "1^128^__NULL__^Examples/Hurricanes",
         path: "Examples/Hurricanes",
         label: "Hurricanes",
         global: true,
         identifier: "id2"
      }
   ];

   beforeEach(() => {
      adapter = {
         get: jest.fn(),
         post: jest.fn(),
         put: jest.fn(),
         delete: jest.fn(),
         map: jest.fn(),
         mapText: jest.fn(),
         mapBinary: jest.fn()
      };
      client = new ViewsheetClient(adapter, () => Promise.resolve("secret"), "http://localhost:8080/sree/");
   });

   describe("when getViewsheets() is called", () => {
      it("then a list of sheets is returned", () => {
         adapter.get.mockReturnValueOnce(Promise.resolve({data: {sheets}}));
         adapter.map.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.getViewsheets().then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/viewsheets");
         });
      });
   });

   describe("when getOpenedViewsheets() is called", () => {
      it("then a list of sheets is returned", () => {
         adapter.get.mockReturnValueOnce(Promise.resolve({data: {sheets: openSheets}}));
         adapter.map.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.getOpenedViewsheets().then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/viewsheets/open");
         });
      });
   });

   describe("when openViewsheet() is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({data: openSheets[0]}));
         adapter.map.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.openViewsheet("1^128^__NULL__^Census", {p1: "v1", p2: "v2"}).then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/viewsheets/open");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when getOpenedViewsheet() is called", () => {
      it("then the opened viewsheet is returned", () => {
         adapter.get.mockReturnValueOnce(Promise.resolve({data: openSheets[0]}));
         adapter.map.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.getOpenedViewsheet("id1").then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/viewsheets/open/id1");
         });
      });
   });

   describe("when closeViewsheet() is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.delete.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);
         return client.closeViewsheet("id1").then(() => {
            expect(adapter.delete.mock.calls.length).toBe(1);
            expect(adapter.delete.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/viewsheets/open/id1");
         });
      });
   });

   describe("when mailViewsheet() is called", () => {
      it("then the adapter is called with the correct arguments", () => {
         adapter.post.mockReturnValueOnce(Promise.resolve({}));
         adapter.map.mockImplementation((response: Promise<any>) => response);
         return client.mailViewsheet("id1", "PDF", ["joe@example.com", "sue@example.com"], "ann@example.com", "Test Mail", "This is a test email.").then(() => {
            expect(adapter.post.mock.calls.length).toBe(1);
            expect(adapter.post.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/viewsheets/open/id1/mail");
            expect(adapter.post.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when exportViewsheet() is called", () => {
      it("then the exported viewsheet file is returned", () => {
         adapter.get.mockReturnValueOnce(Promise.resolve({data: createBinaryData()}));
         adapter.mapBinary.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.exportViewsheet("id1", "PDF").then(result => {
            validateBinaryData(result);
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/viewsheets/open/id1/export/PDF");
         });
      });
   });
});

function createBinaryData(): ArrayBuffer {
   const buffer = new ArrayBuffer(16);
   const array = new Int32Array(buffer);
   array[0] = 1;
   array[1] = 2;
   array[2] = 3;
   array[3] = 4;
   return buffer;
}

function validateBinaryData(buffer: ArrayBuffer): void {
   expect(buffer).toBeTruthy();
   expect(buffer.byteLength).toBe(16);
   const actual = new Int32Array(buffer);
   expect(actual[0]).toBe(1);
   expect(actual[1]).toBe(2);
   expect(actual[2]).toBe(3);
   expect(actual[3]).toBe(4);
}
