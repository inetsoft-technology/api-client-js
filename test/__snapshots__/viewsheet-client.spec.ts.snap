// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewsheetClient when getOpenedViewsheet() is called then the opened viewsheet is returned 1`] = `
Object {
  "asset": "1^128^__NULL__^Examples/Census",
  "global": true,
  "identifier": "id1",
  "label": "Census",
  "path": "Examples/Census",
}
`;

exports[`ViewsheetClient when getOpenedViewsheets() is called then a list of sheets is returned 1`] = `
Array [
  Object {
    "asset": "1^128^__NULL__^Examples/Census",
    "global": true,
    "identifier": "id1",
    "label": "Census",
    "path": "Examples/Census",
  },
  Object {
    "asset": "1^128^__NULL__^Examples/Hurricanes",
    "global": true,
    "identifier": "id2",
    "label": "Hurricanes",
    "path": "Examples/Hurricanes",
  },
]
`;

exports[`ViewsheetClient when getViewsheets() is called then a list of sheets is returned 1`] = `
Array [
  Object {
    "asset": "1^128^__NULL__^Examples/Census",
    "global": true,
    "label": "Census",
    "path": "Examples/Census",
  },
  Object {
    "asset": "1^128^__NULL__^Examples/Hurricanes",
    "global": true,
    "label": "Hurricanes",
    "path": "Examples/Hurricanes",
  },
  Object {
    "asset": "1^128^__NULL__^Examples/Sales Summary",
    "global": true,
    "label": "Sales Summary",
    "path": "Examples/Sales Summary",
  },
]
`;

exports[`ViewsheetClient when mailViewsheet() is called then the adapter is called with the correct arguments 1`] = `
Object {
  "format": "PDF",
  "message": "This is a test email.",
  "recipients": Array [
    "joe@example.com",
    "sue@example.com",
  ],
  "sender": "ann@example.com",
  "subject": "Test Mail",
}
`;

exports[`ViewsheetClient when openViewsheet() is called then the adapter is called with the correct arguments 1`] = `
Object {
  "asset": "1^128^__NULL__^Examples/Census",
  "global": true,
  "identifier": "id1",
  "label": "Census",
  "path": "Examples/Census",
}
`;

exports[`ViewsheetClient when openViewsheet() is called then the adapter is called with the correct arguments 2`] = `
Object {
  "asset": "1^128^__NULL__^Census",
  "parameters": Object {
    "p1": "v1",
    "p2": "v2",
  },
}
`;
